src/sendCommercialMetrics.ts

```diff
@@ -1,72 +1,186 @@
+};
+
+const getDevProperties = (isDev: boolean): Property[] => {
+	return isDev ? [{ name: 'isDev', value: window.location.hostname }] : [];
+};
+
+const getAdBlockerProperties = (adBlockerInUse?: boolean): Property[] => {
+	return adBlockerInUse !== undefined
+		? [
+				{
+					name: 'adBlockerInUse',
+					value: adBlockerInUse.toString(),
+				},
+		  ]
+		: [];
+};
+
+const filterUndefinedEventTimerProperties = (
+	eventTimerProperties: EventProperties,
+): Array<[string, string | number]> => {
+	// Transforms object {key: value} pairs into an array of [key, value] arrays
+	return Object.entries(eventTimerProperties).filter(
+		([, value]) => typeof value !== 'undefined',
+	);
+};
+
+const mapEventTimerPropertiesToString = (
+	properties: Array<[string, string | number]>,
+): Property[] => {
+	return properties.map(([name, value]) => ({
+		name: String(name),
+		value: String(value),
+	}));
+};
+
+const roundTimeStamp = (events: TimedEvent[]): Metric[] => {
+	return events.map(({ name, ts }) => ({
+		name,
+		value: Math.ceil(ts),
+	}));
 };
 
-export function sendCommercialMetrics(
+function sendMetrics(isDev: boolean) {
+	const endpoint = getEndpoint(isDev);
+
+	log(
+		'commercial',
+		'About to send commercial metrics',
+		commercialMetricsPayload,
+	);
+	return navigator.sendBeacon(
+		endpoint,
+		JSON.stringify(commercialMetricsPayload),
+	);
+}
+
+/**
+ * A method to asynchronously send metrics after initialization.
+ * @param init.isDev - used to determine whether to use CODE or PROD endpoints.
+ */
+export function bypassCommercialMetricsSampling(isDev: boolean): void {
+	if (!initialised) {
+		console.warn('initCommercialMetrics not yet initialised');
+		return;
+	}
+
+	sendMetrics(isDev);
+}
+
+/**
+ * A method to initialise metrics.
+ * @param init.pageViewId - identifies the page view. Usually available on `guardian.config.ophan.pageViewId`. Defaults to `null`
+ * @param init.browserId - identifies the browser. Usually available via `getCookie({ name: 'bwid' })`. Defaults to `null`
+ * @param init.isDev - used to determine whether to use CODE or PROD endpoints.
+ * @param init.adBlockerInUse - indicates whether or not ann adblocker is being used.
+ */
+export function initCommercialMetrics(
 	pageViewId: string,
 	browserId: string | undefined,
 	isDev: boolean,
 	adBlockerInUse?: boolean,
 ): boolean {
-	const devProperties: Properties[] = isDev
-		? [{ name: 'isDev', value: window.location.hostname }]
-		: [];
-	const adBlockerProperties: Properties[] =
-		adBlockerInUse !== undefined
-			? [
-					{
-						name: 'adBlockerInUse',
-						value: adBlockerInUse.toString(),
-					},
-			  ]
-			: [];
-
-	const endpoint = isDev
-		? '//performance-events.code.dev-guardianapis.com/commercial-metrics'
-		: '//performance-events.guardianapis.com/commercial-metrics';
-	if (document.visibilityState !== 'hidden') return false;
+	if (document.visibilityState !== 'hidden') {
+		return false;
+	}
```

Comment:
This should be wrapped inside a `visibilitychange` event listener.

Reply:
👍🏻